### PR TITLE
L1D cache optiizations

### PR DESCRIPTION
**Pull request by MajidAbdelilah**
this is the L1D cache optimization pull request 
in this request i have restructured the vertex and triangle structure to be blocks of 8 values

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,6 +8,15 @@ message(STATUS "Building with: ${CMAKE_CXX_COMPILER_ID}")
...
 
+if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+	set(CMAKE_CXX_FLAGS "-g")
+	set(CMAKE_C_FLAGS "-g")
+elseif(CMAKE_BUILD_TYPE STREQUAL "Release")
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Ofast -march=native -mtune=native -ftree-loop-distribution -floop-interchange -freorder-blocks-and-partition ")
+endif()
+
+
+
```
Comment (VonTum) [inline CMakeLists.txt:19]: I see you have edited the CMakeLists.txt. We actually already add debug symbols, if you look at the lines below `if (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")`
GCC specific Debug and Release flags can be added below the `else()`
By default Release builds already have optimization flag -O3, we already define -ffast-math so -Ofast is redundant. I'm not sure about the other optimization flags but I believe they're included in -O3?

File: Physics3D/geometry/triangleMesh.cpp [hunk 2]
```diff
@@ -33,15 +37,15 @@ static UniqueAlignedPointer<T> copy(const UniqueAlignedPointer<T>& buf, size_t s
...
+	size_t offset = ((size + 7) & 0xFFFFFFFFFFFFFFF8);
 	T* xValues = buf;
 	T* yValues = buf + offset;
 	T* zValues = buf + 2 * offset;
 
 	for(size_t i = size; i < offset; i++) {
 		xValues[i] = xValues[size - 1];
 		yValues[i] = yValues[size - 1];
-		zValues[i] = zValues[size - 1];
+                zValues[i] = zValues[size - 1];
```
Comment (VonTum) [inline Physics3D/geometry/triangleMesh.cpp:48]: Inconsistent indentation

File: Physics3D/geometry/triangleMesh.cpp [hunk 3]
```diff
@@ -136,35 +140,35 @@ EditableMesh::EditableMesh(MeshPrototype&& mesh) noexcept : MeshPrototype(std::m
...
+	this->vertices[correct_index + 8] = newVertex.y;
+	this->vertices[correct_index + 2 * 8] = newVertex.z;
 }
 void EditableMesh::setVertex(int index, float x, float y, float z) {
 	assert(index >= 0 && index < vertexCount);
-	size_t offset = getOffset(vertexCount);
-	this->vertices[index] = x;
-	this->vertices[index + offset] = y;
-	this->vertices[index + 2 * offset] = z;
+	size_t correct_index = (index/8)*8*2+index; //getOffset(vertexCount);
```
Comment (VonTum) [inline Physics3D/geometry/triangleMesh.cpp:150]: Minor nitpicks on the coding style. 
our variables are formatted in camelCase, also, I'd add a comment explaining the *8*2+index, it threw me for a loop before I understood.

File: Physics3D/geometry/triangleMesh.cpp [hunk 4]
```diff
@@ -177,31 +181,82 @@ TriangleMesh::TriangleMesh(UniqueAlignedPointer<float>&& vertices, UniqueAligned
...
-	size_t vertexOffset = getOffset(vertexCount);
+        size_t vertexOffset = 8;//getOffset(vertexCount);
 
+	
+	//printf("vertex_count = ori:%d, now: %lu\ntriangle_count = ori:%d, now: %lu\n", vertexCount, getOffset(vertexCount) * 3, triangleCount, getOffset(triangleCount) * 3),
+	
+	//printf("TriangleMesh::TriangleMesh->vertex_offset = %lu\n", vertexOffset);
+	
+	//printf("TriangleMesh::TriangleMesh->vertexCount = %d\n", vertexCount);
+
```
Comment (VonTum) [inline Physics3D/geometry/triangleMesh.cpp:192]: This file is filled with commented out debug prints, remove. Also inconsistent indentation

File: Physics3D/geometry/triangleMesh.cpp [hunk 5]
```diff
@@ -177,31 +181,82 @@ TriangleMesh::TriangleMesh(UniqueAlignedPointer<float>&& vertices, UniqueAligned
...
-		yValues[i] = vertices[i].y;
-		zValues[i] = vertices[i].z;
+	    xValues[a] = vertices[i].x;
+	    yValues[a] = vertices[i].y;
+	    zValues[a] = vertices[i].z;
+	    /*  printf("x = %f\n", xValues[a]);
+	    printf("y = %f\n", yValues[a]);
+	    printf("z = %f\n", zValues[a]);
+	    */a++;
+	    a+=vertexOffset*((a%vertexOffset)==0)*2;
```
Comment (VonTum) [inline Physics3D/geometry/triangleMesh.cpp:206]: Minor style nitpick, add spacing around operators. `vertexOffset * ((a % vertexOffset) == 0) * 2`

Also, what is 'a'?

File: Physics3D/geometry/triangleMesh.cpp [hunk 6]
```diff
@@ -177,31 +181,82 @@ TriangleMesh::TriangleMesh(UniqueAlignedPointer<float>&& vertices, UniqueAligned
...
+	    //printf("a = %u\n", a);
+	    
 	}
-	fixFinalBlock(this->vertices.get(), vertexCount);
 
-	size_t triangleOffset = getOffset(triangleCount);
+	unsigned int last_index = a-1;
+	unsigned int size_left =  (((vertexCount + 7) & 0xFFFFFFFFFFFFFFF8)*3-vertexCount*3)/3;
```
Comment (VonTum) [inline Physics3D/geometry/triangleMesh.cpp:212]: Lots of math on a single line, split computation over multiple lines with named variables
Also, camelCase names. 
Lastly, the constant `& 0xFFFFFFFFFFFFFFF8` is 64 bits wide. unsigned int is (usually) 32 bits wide. Instead I recommend: `& ~0x7`

File: Physics3D/geometry/triangleMesh.cpp [hunk 7]
```diff
@@ -177,31 +181,82 @@ TriangleMesh::TriangleMesh(UniqueAlignedPointer<float>&& vertices, UniqueAligned
...
+		printf("c = %d\n", cValues[a]);
+		*/a++;
+		a+=triangleOffset*((a%triangleOffset)==0)*2;
+	        //printf("a_index = %u\n", a);
 	}
-	fixFinalBlock(this->triangles.get(), triangleCount);
+
+	
+	last_index = a-1;
+	size_left =  ((((triangleCount + 7) & 0xFFFFFFFFFFFFFFF8)*3-triangleCount*3)/3);
```
Comment (VonTum) [inline Physics3D/geometry/triangleMesh.cpp:247]: Don't reuse the same variable. Create a new variable. This code could actually be extracted into a static function

File: Physics3D/geometry/triangleMesh.cpp [hunk 8]
```diff
@@ -221,10 +276,16 @@ TriangleMesh::TriangleMesh(MeshPrototype&& mesh) noexcept :
 }
 
 IteratorFactory<ShapeVertexIter> TriangleMesh::iterVertices() const {
-	return IteratorFactory<ShapeVertexIter>(ShapeVertexIter{vertices, getOffset(vertexCount)}, ShapeVertexIter{vertices + vertexCount, getOffset(vertexCount)});
+  return IteratorFactory<ShapeVertexIter>(ShapeVertexIter{vertices, 8, 0},
+					  ShapeVertexIter{&vertices[getOffset(vertexCount) * 3 -
+								    ((getOffset(vertexCount)*3-vertexCount*3)/3) - 8*2 -
+								    1 * (vertexCount%8==0)], 8, 0});
```
Comment (VonTum) [inline Physics3D/geometry/triangleMesh.cpp:282]: I find the math a bit hard to follow. 
It might be best to add some intermediate variables to explain the math

**Comment by olympus112** [comment]
Hi, thanks for the pull request!

Sadly, we can not allow a part of this pull request: your modification (and deletion) of instance.shader is not a real fix for your problem and will remove the ability to use texture batching. The instance shader (together with basic.shader) is the main shader for the application and thus can not be removed. If you have a problem, you can make an issue for it and I will give it a look.

Furthermore, it would be nice if you could remove debug prints and comments from your commits, it looks cleaner that way.

File: Physics3D/geometry/triangleMeshAVX.cpp [hunk 9]
```diff
@@ -73,29 +73,30 @@ Vec3f TriangleMesh::furthestInDirectionAVX(const Vec3f& direction) const {
...
+	  __m256 zVal = _mm256_load_ps(zValues + blockI * 24);
+	  
+	  __m256 dot = _mm256_fmadd_ps(dz, zVal, _mm256_fmadd_ps(dy, yVal, _mm256_mul_ps(dx, xVal))); 
+	  
+	  __m256 whichAreMax = _mm256_cmp_ps(dot, bestDot, _CMP_LT_OQ); // Greater than, false if dot == NaN
+	
+	  bestDot = _mm256_blendv_ps(dot, bestDot, whichAreMax);
+	  bestX = _mm256_blendv_ps(xVal, bestX, whichAreMax);
+	  bestY = _mm256_blendv_ps(yVal, bestY, whichAreMax);
+	  bestZ = _mm256_blendv_ps(zVal, bestZ, whichAreMax);
```
Comment (VonTum) [inline Physics3D/geometry/triangleMeshAVX.cpp:99]: Looks great! The code above with all commented out debug prints should look more like this.

File: Physics3D/misc/validityHelper.cpp [hunk 10]
```diff
@@ -76,7 +76,7 @@ bool isValid(const TriangleMesh& mesh) {
 	for(int i = 0; i < mesh.triangleCount; i++) {
 		const Triangle& t = mesh.getTriangle(i);
 		if(!isValidTriangle(t, mesh.vertexCount)) {
-			Debug::logError("Invalid triangle! Triangle %d {%d, %d, %d} points to a nonexistent vertex or has a duplicate index!", i, t.firstIndex, t.secondIndex, t.thirdIndex);
+			Debug::logError("Invalid triangle! Triangle %d {%d, %d, %d} points to a nonexistent vertex or has a duplicate index!\n", i, t.firstIndex, t.secondIndex, t.thirdIndex);
```
Comment (VonTum) [inline Physics3D/misc/validityHelper.cpp:79]: Nice catch

**Review by VonTum — CHANGES_REQUESTED** [review]
Alright, so first of all from just looking at the code I have a few observations:
The first thing I notice is the code is indeed *very* messy:
- Lots of commented-out code
- Debug prints
- Inconsistent indentation: We use tabs instead of spaces. Configure your editor to use tabs. 
- Minor, but inconsistent operator spacing. Try to stick to the existing style. This is spaces around every operator
- Finally, Spelling: "cach-opt", "optiization". 

I see you have edited the CMakeLists.txt. We actually already add debug symbols, if you look at the lines below `if (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")`
GCC specific Debug and Release flags can be added below the `else()`

I find the math a bit hard to follow. 
`a+=vertexOffset*((a%vertexOffset)==0)*2;` is this to advance to the next block? 
Similarly these lines:
```c++
ShapeTriangleIter{&triangles[getOffset(triangleCount) *3 -
    ((getOffset(triangleCount)*3-triangleCount*3)/3) - 8*2 -
        1 * (triangleCount%8==0)], 8, 0});
```
It might be best to add some intermediate variables to explain the math

**Comment by MajidAbdelilah** [comment]
@olympus112  sorry for deleting the shaders i dont use github that much espicialy for pull requests i thought by hitting the delete button it will remove it only from the pull request and not deleting the actual files
